File: internal/vision/roboflow.go
package vision

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"mime/multipart"
	"net/http"
	"roborok/internal/common"
	"time"
)

// RoboflowResponse represents the response from Roboflow API
type RoboflowResponse struct {
	Predictions []struct {
		X          float64 `json:"x"`
		Y          float64 `json:"y"`
		Width      float64 `json:"width"`
		Height     float64 `json:"height"`
		Confidence float64 `json:"confidence"`
		Class      string  `json:"class"`
	} `json:"predictions"`
	Time  float64 `json:"time"`
	Image struct {
		Width  int `json:"width"`
		Height int `json:"height"`
	} `json:"image"`
}

// SendToRoboflow sends an image to Roboflow for inference
func SendToRoboflow(imageBytes []byte, apiKey, modelID string) (*RoboflowResponse, error) {
	// Add detailed logging before the API call
	// log.Printf("Sending request to Roboflow API - Model ID: %s", modelID)

	// Safely log API key (just first few chars)
	if len(apiKey) > 4 {
		// log.Printf("API Key (first 4 chars): %s...", apiKey[:4]) // Only log first 4 chars for security
	} else {
		log.Printf("API Key: [TOO SHORT - POSSIBLE ERROR]")
	}

	url := fmt.Sprintf("https://detect.roboflow.com/%s?api_key=%s", modelID, apiKey)

	// Create multipart form
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	// Add the file
	part, err := writer.CreateFormFile("file", "screenshot.png")
	if err != nil {
		return nil, fmt.Errorf("failed to create form file: %w", err)
	}

	_, err = io.Copy(part, bytes.NewReader(imageBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to copy image data: %w", err)
	}

	writer.Close()

	// Create and send the request
	req, err := http.NewRequest("POST", url, body)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", writer.FormDataContentType())

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// Check for success with more detailed error reporting
	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		log.Printf("Roboflow API error - Status: %d, Response: %s", resp.StatusCode, string(bodyBytes))

		// Try to parse the error message for more details
		var errorResponse struct {
			Message string `json:"message"`
			Detail  string `json:"detail,omitempty"`
		}

		if err := json.Unmarshal(bodyBytes, &errorResponse); err == nil && errorResponse.Detail != "" {
			return nil, fmt.Errorf("API error (status %d): %s - %s",
				resp.StatusCode, errorResponse.Message, errorResponse.Detail)
		}

		return nil, fmt.Errorf("API error (status %d): %s", resp.StatusCode, string(bodyBytes))
	}

	// Parse the response
	var result RoboflowResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	// Log detailed information about detected objects
	if len(result.Predictions) > 0 {
		log.Printf("Roboflow API request successful - Detected %d objects:", len(result.Predictions))
		for i, pred := range result.Predictions {
			log.Printf("  %d. %s (%.2f): (%.1f, %.1f) %.0fx%.0f",
				i+1, pred.Class, pred.Confidence, pred.X, pred.Y, pred.Width, pred.Height)
		}
	} else {
		log.Printf("Roboflow API request successful - No objects detected")
	}

	return &result, nil
}

// AnalyzeGameState analyzes the current game state and view
func AnalyzeGameState(
	screenshot []byte,
	apiKey string,
	modelID string,
) (string, []common.Detection, error) {
	// Send screenshot to Roboflow
	resp, err := SendToRoboflow(screenshot, apiKey, modelID)
	if err != nil {
		return "", nil, fmt.Errorf("failed to analyze screenshot: %w", err)
	}

	// Convert response to detections
	var detections []common.Detection
	for _, pred := range resp.Predictions {
		detections = append(detections, common.Detection{
			Class:      pred.Class,
			X:          pred.X,
			Y:          pred.Y,
			Width:      pred.Width,
			Height:     pred.Height,
			Confidence: pred.Confidence,
		})
	}

	// Determine the view (city or map)
	gameView := DetermineGameView(detections)

	return gameView, detections, nil
}

// DetermineGameView determines if we're in city view, map view, or unknown view
func DetermineGameView(detections []common.Detection) string {
	// First, check explicit view indicators
	for _, detection := range detections {
		if detection.Class == "on_field" && detection.Confidence > common.MinConfidence {
			return "field"
		}

		if detection.Class == "in_city" && detection.Confidence > common.MinConfidence {
			return "city"
		}
	}

	// If no explicit indicator, check for view-specific elements
	cityIndicators := 0
	mapIndicators := 0

	for _, detection := range detections {
		// City view indicators
		if detection.Class == "city_hall" ||
			detection.Class == "city_hall_upgradeable" ||
			detection.Class == "barracks" ||
			detection.Class == "barracks_upgradeable" ||
			detection.Class == "barracks_upgradeable_idle" ||
			detection.Class == "farm" ||
			detection.Class == "builders_hut" ||
			detection.Class == "builders_hut_idle" ||
			detection.Class == "tavern" ||
			detection.Class == "tavern_upgradeable_clickable" {
			cityIndicators++
		}

		// Map view indicators
		if detection.Class == "return_to_city_button" ||
			detection.Class == "world_map" ||
			detection.Class == "barbarian" ||
			detection.Class == "resource_node" {
			mapIndicators++
		}
	}

	// Determine view based on the count of indicators
	if cityIndicators > 0 && cityIndicators > mapIndicators {
		return "city"
	} else if mapIndicators > 0 {
		return "map"
	}

	// Default to unknown if we can't determine
	return "city"
}

File: roborok/vision/roboflow.py
# roborok/roboflow.py
"""Roboflow API integration"""

import requests
import json
import logging
import time
from typing import List, Dict, Any, Optional

from roborok.models import Detection
from roborok.vision.screenshot import capture_screenshot
from roborok.utils.adb import tap_screen, swipe_screen

def send_to_roboflow(image_bytes, api_key, model_id):
    """
    Send an image to Roboflow for detection
    
    Args:
        image_bytes: Screenshot image data as bytes
        api_key: Roboflow API key
        model_id: Roboflow model ID (format: project_name/version_number)
        
    Returns:
        Roboflow API response JSON if successful, None if failed
    """
    url = f"https://detect.roboflow.com/{model_id}?api_key={api_key}"
    
    try:
        # Create files dict for the multipart request
        files = {'file': ('screenshot.jpg', image_bytes, 'image/jpeg')}
        
        # Send the request
        response = requests.post(url, files=files, timeout=10)
        
        # Raise an exception for bad status codes
        response.raise_for_status()
        
        # Parse JSON response
        return response.json()
    except Exception as e:
        print(f"Error sending image to Roboflow: {e}")
        return None

def parse_detections(roboflow_response):
    """
    Parse Roboflow API response into Detection objects
    
    Args:
        roboflow_response: JSON response from Roboflow API
        
    Returns:
        List of Detection objects
    """
    detections = []
    
    if not roboflow_response or 'predictions' not in roboflow_response:
        return detections
    
    # Parse each prediction into a Detection object
    for pred in roboflow_response['predictions']:
        detection = Detection(
            class_name=pred['class'],
            x=pred['x'],
            y=pred['y'],
            width=pred['width'],
            height=pred['height'],
            confidence=pred['confidence']
        )
        detections.append(detection)
    
    return detections

def detect_game_elements(screenshot_bytes, api_key, model_id):
    """
    Process a screenshot and detect game elements
    
    Args:
        screenshot_bytes: Screenshot image data
        api_key: Roboflow API key
        model_id: Roboflow model ID
        
    Returns:
        Tuple of (List[Detection], error_message)
    """
    # Send to Roboflow
    response = send_to_roboflow(screenshot_bytes, api_key, model_id)
    
    if not response:
        return [], "Failed to get response from Roboflow"
    
    # Parse detections
    detections = parse_detections(response)
    
    # Log detection count
    print(f"Detected {len(detections)} game elements")
    
    return detections, None

def determine_game_view(detections):
    """
    Determine if we're in city view, map view, etc.
    
    Args:
        detections: List of Detection objects
        
    Returns:
        String indicating the view ("city", "map", "unknown")
    """
    # Check for explicit view indicators
    for detection in detections:
        if detection.class_name == "in_city" and detection.is_confident():
            return "city"
        if detection.class_name == "on_map" and detection.is_confident():
            return "map"
    
    # Check for view-specific elements
    city_indicators = 0
    map_indicators = 0
    
    for detection in detections:
        # City view indicators
        if detection.class_name in [
            "city_hall", "barracks", "farm", "storehouse"
        ] and detection.is_confident():
            city_indicators += 1
        
        # Map view indicators
        if detection.class_name in [
            "return_to_city_button", "map_button", "barbarian"
        ] and detection.is_confident():
            map_indicators += 1
    
    # Determine view based on indicators
    if city_indicators > map_indicators:
        return "city"
    elif map_indicators > 0:
        return "map"
    else:
        return "unknown"

def capture_and_detect(device_id: str, adb_path: str, api_key: str, model_id: str, max_attempts: int = 6) -> List[Detection]:
    """
    Take a screenshot, handle all high-priority clickables in order of priority, and return clean detections.
    
    Args:
        device_id: Device ID
        adb_path: Path to ADB executable
        api_key: Roboflow API key
        model_id: Roboflow model ID
        max_attempts: Maximum number of attempts to handle high-priority elements
        
    Returns:
        List of detections after handling all high-priority elements
    """
    logger = logging.getLogger("actions.capture_and_detect")
    
    # Define high-priority clickable elements (in order of priority)
    high_priority_elements = [
        "help_chat_bubble",  # Handle help bubbles first as they block interaction
        "farm_clickable", 
        "lumber_mill_clickable",
        "alliance_help_request",
        # "alliance_invite_join",  # Removed from direct click list - will handle specially
        # Add more elements here in order of priority
    ]
    
    for attempt in range(max_attempts):
        # Take a screenshot
        screenshot_bytes = capture_screenshot(device_id, adb_path)
        if not screenshot_bytes:
            logger.error(f"Failed to capture screenshot (attempt {attempt+1}/{max_attempts})")
            time.sleep(0.5)
            continue
        
        # Get detections
        detections, error = detect_game_elements(screenshot_bytes, api_key, model_id)
        if error:
            logger.error(f"Error getting detections (attempt {attempt+1}/{max_attempts}): {error}")
            time.sleep(0.5)
            continue
        
        # Special case: Check for alliance invite dialog
        has_alliance_invite = False
        has_exit_button = False
        exit_button = None
        
        for det in detections:
            if det.class_name == "alliance_invite_join" and det.is_confident(0.4):
                has_alliance_invite = True
            elif det.class_name == "exit_dialog_button" and det.is_confident(0.4):
                has_exit_button = True
                exit_button = det
        
        # If we have both alliance invite and exit button, click the exit button
        if has_alliance_invite and has_exit_button and exit_button is not None:
            logger.info(f"Found alliance invite dialog, clicking exit button at ({exit_button.x}, {exit_button.y})")
            tap_screen(device_id, adb_path, int(exit_button.x), int(exit_button.y))
            time.sleep(0.5)  # Wait for action to register
            continue  # Get fresh detections
            
        # Check for other high-priority clickables in order of priority
        high_priority_found = False
        
        # First, collect all high-priority elements present
        priority_detections = []
        for priority_class in high_priority_elements:
            for det in detections:
                if det.class_name == priority_class and det.is_confident(0.4):
                    priority_detections.append(det)
        
        # If we found any high-priority elements, click the highest priority one
        if priority_detections:
            # Sort by priority (order in the high_priority_elements list)
            sorted_priorities = sorted(
                priority_detections, 
                key=lambda det: high_priority_elements.index(det.class_name)
            )
            
            # Click the highest priority element
            top_priority = sorted_priorities[0]
            logger.info(f"Found high-priority element {top_priority.class_name} at ({top_priority.x}, {top_priority.y}), clicking")
            tap_screen(device_id, adb_path, int(top_priority.x), int(top_priority.y))
            time.sleep(0.5)  # Wait for action to register
            
            # If we found and handled a high-priority element, try again to get fresh detections
            high_priority_found = True
        
        if high_priority_found:
            logger.info("Handled high-priority element, getting fresh detections")
            continue
            
        # If no high-priority elements found, return the detections
        return detections
    
    # If we've exhausted all attempts
    logger.info(f"Completed {max_attempts} rounds of checking for high-priority elements")
    return detections